### DISCUSSION TITLE
yarn upgrade known issues

### DISCUSSION THREAD

**Discussion by dvh91** [renovatebot/renovate#8092] [Request Help]
Following yarn issues https://github.com/yarnpkg/yarn/issues/3967 and https://github.com/yarnpkg/yarn/issues/3202, I was wondering if `yarn upgrade` command can be replaced with `yarn remove dep` and then `yarn add dep@version` to avoid these bugs.

Suggested diff in [here](https://github.com/renovatebot/renovate/blob/master/lib/manager/npm/post-update/yarn.ts#L128):

```diff
- commands.push(
-   `yarn upgrade ${lockUpdates
-     .map((update) => update.depName)
-     .join(' ')} ${cmdOptions}`.trim()
- );

+ commands.push(
+   `yarn remove ${lockUpdates
+     .map((update) => update.depName)
+     .join(' ')}`.trim()
+ );
+ commands.push(
+   `yarn add ${lockUpdates
+     .map((update) => update.depName)
+     .join(' ')} ${cmdOptions}`.trim()
+ );
```

**Comment by rarkins** [comment]
Will that always produce the correct result though? e.g. let's say the dependency is defined as ^2.0.0 but that latest version is 3.1.0. If you run "yarn upgrade" it will limit the upgrade to `^2.0.0` however if you "yarn add" then won't it install 3.1.0?

**Reply by dvh91** [reply]
If you'll `yarn add dep@^2.0.0` then it will produce the same result.

**Reply by rarkins** [reply]
OK, please raise a PR with the appropriate changes and we'll test it out. When multiple dependencies are updated, do you expect there to be any difference between remove-remove-removate-add-add-add vs add-remove-add-remove-add-remove ?
